Dropp/modules/firebase_mod.js

```diff
@@ -11,31 +12,360 @@ admin.initializeApp({
 });
 
 db = admin.database();
-
 var exports = module.exports = {};
 
-exports.test = function() {
-	console.log('test');
-}
+// Validates username and password from request with database record
+exports.validateCredentials = function(username, password, callback) {
+	// Declare error variables
+	var dbReason, errorType, explanation;
+
+	// Retrieve password for requested user
+	var passwordRef = db.ref('/passwords/' + username);
+	passwordRef.once('value', snapshot => {
+		if (snapshot.val() == null) {
+			dbReason    = username + ' does not exist in passwords table';
+			errorType   = 'login_error';
+			explanation = 'The username or password is incorrect';
+			return callback(fail('validateCredentials', dbReason, errorType, explanation));
+		}
+
+		// Compare the entered password to the database password
+		bcrypt.compare(password, snapshot.val(), (err, res) => {
+			if (err) {
+        switch (err) {
+					case 'Not a valid BCrypt hash.':
+						dbReason    = 'Password for ' + username + ' is not hashed in the database';
+						errorType   = 'api_error';
+						explanation = 'There was a problem with our back-end services';
+						break;
+          default:
+            dbReason    = 'unknown';
+            errorType   = 'api_error';
+            explanation = 'There was a problem with our back-end services';
+        }
+
+        return callback(fail('validateCredentials', dbReason, errorType, explanation));
+      }
+
+			if (!res) {
+        dbReason    = password + ' does not match ' + username + '\'s password';
+        errorType   = 'login_error';
+        explanation = 'The username or password is incorrect';
+        return callback(fail('validateCredentials', dbReason, errorType, explanation));
+      }
+
+			var successJson = {
+        success: {
+          message: 'Valid login credentials',
+        }
+      };
+
+      callback(successJson);
+		});
+	}, err => {
+		// Determine type of error
+		var errorCode = -1; // FIXME: Parse err object for more specific information
+		switch (errorCode) {
+			default:
+				dbReason    = err;
+				errorType   = 'api_error';
+				explanation = 'There was a problem with our back-end services';
+		}
+
+		return callback(fail('validateCredentials', dbReason, errorType, explanation));
```

Comment:
callback is necessary for asynchronous performance and correct values being returned from async firebase calls